src/commands/git-metadata/gitdb.ts

```diff
@@ -0,0 +1,297 @@
+    await git.addConfig('remote.origin.partialclonefilter', 'blob:none')
+    log.info('[unshallow] Running git fetch --shallow-since="1 month ago" --update-shallow --refetch')
+    await git.fetch(['--shallow-since="1 month ago"', '--update-shallow', '--refetch'])
+    log.info('[unshallow] Fetch completed.')
+  }
+}
+
+// getKnownCommits asks the backend which of the given commits are already known
+const getKnownCommits = async (log: Logger, request: RequestBuilder, repoURL: string, latestCommits: string[]) => {
+  interface SearchCommitResponse {
+    data: Commit[]
+  }
+
+  interface Commit {
+    type: string
+    id: string
+  }
+
+  const localCommitData = JSON.stringify({
+    meta: {
+      repository_url: repoURL,
+    },
+    data: latestCommits.map((commit) => ({
+      id: commit,
+      type: 'commit',
+    })),
+  })
+  const response = await runRequest(log, 'search_commits', () =>
+    request({
+      url: '/api/v2/git/repository/search_commits',
+      headers: {
+        'Content-Type': 'application/json',
+      },
+      data: localCommitData,
+      method: 'POST',
+      timeout: API_TIMEOUT,
+    })
+  )
+  const commits = response.data as SearchCommitResponse
+  if (!commits || commits.data === undefined) {
+    throw new Error(`Invalid API response: ${response}`)
+  }
+
+  return commits.data.map((c) => {
+    if (c.type !== 'commit' || c.id === undefined) {
+      throw new Error('Invalid commit type response')
+    }
+
+    return sanitizeCommit(c.id)
+  })
+}
+
+const sanitizeCommit = (sha: string) => {
+  const isValidSha = (s: string) => /[0-9a-f]{40}/.test(s)
+
+  const sanitizedCommit = sha.replace(/[^0-9a-f]+/g, '')
+  if (sanitizedCommit !== sha) {
+    throw new Error(`Invalid commit format: ${sha} (different from sanitized ${sanitizedCommit})`)
+  }
+  if (!isValidSha(sanitizedCommit)) {
+    throw new Error(`Invalid commit format: ${sanitizedCommit}`)
+  }
+
+  return sanitizedCommit
+}
+
+const getObjectsToUpload = async (git: simpleGit.SimpleGit, commitsToExclude: string[]) => {
+  const rawResponse = await git.raw(
+    ['rev-list', '--objects', '--no-object-names', '--filter=blob:none', '--since="1 month ago"', 'HEAD'].concat(
+      commitsToExclude.map((sha) => '^' + sha)
+    )
+  )
+  const commitsToInclude = rawResponse.split('\n').filter((c) => c !== '')
+
+  return commitsToInclude
+}
+
+const generatePackFilesForCommits = (log: Logger, commits: string[]) => {
+  if (commits.length <= 0) {
+    return []
+  }
+
+  const generatePackfiles = (packfilePath: string) => {
+    const packObjectResults = child_process
+      .execSync(`git pack-objects --compression=9 --max-pack-size=3m ${packfilePath}`, {
+        input: commits.join('\n'),
+      })
+      .toString()
+      .split('\n')
+      .filter((sha) => sha)
+      .map((sha) => `${packfilePath}-${sha}.pack`)
+
+    return packObjectResults
+  }
+
+  // Try using tmp folder first:
+  try {
+    const tmpFolder = os.tmpdir()
```

Comment:
Why we generate some random prefix (and don't check possible collisions) instead of using something like mkdtemp?

Reply:
This is based on the specs written for all tracers. mkdtemp might be more relevant indeed but the existing implem seems enough for our use case: collisions are still not very likely and wouldn't be a big issue. Most of the time these files will be in a tmpdir anyway, and for the cases where it's in cwd: I don't expect collision to have an impact except for making troubleshooting harder. We're talking about an edge case of an edge case here.

Reply:
Could you please give a link to this specs

Reply:
@sashacmc I reworked this to use `mkdtemp` as well as adding a manual deletion step of the temp folder after the command ran (redundant with the OS in case we're in tmpdir but useful when we are in cwddir).